packages/headless/src/controllers/core/facets/facet-dependencies-manager/headless-facet-dependencies-manager.ts

```diff
@@ -0,0 +1,160 @@
+  FacetOptionsSection,
+  FacetSection,
+  NumericFacetSection,
+} from '../../../../state/state-sections';
+import {loadReducerError} from '../../../../utils/errors';
+
+export interface AnyFacetDependency<T extends AnyFacetValueRequest> {
+  parentFacetId: string;
+  isDependencyMet: (parentValues: T[]) => boolean;
+}
+
+export interface FacetDependenciesManagerProps {
+  dependentFacetId: string;
+  dependencies: AnyFacetDependency<AnyFacetValueRequest>[];
+}
+
+export interface FacetDependenciesManager {
+  stopWatching(): void;
+}
+
+export function buildFacetDependenciesManager(
+  engine: CoreEngine,
+  props: FacetDependenciesManagerProps
+): FacetDependenciesManager {
+  if (!loadFacetDependenciesManagerReducers(engine)) {
+    throw loadReducerError;
+  }
+  const lastObservedValuesPerParentFacet: Record<string, string> = {};
+
+  const isFacetEnabled = (facetId: string) => {
+    return engine.state.facetOptions.facets[facetId]?.enabled ?? true;
+  };
+
+  const getFacetValuesById = (facetId: string) => {
+    if (engine.state.facetSet && facetId in engine.state.facetSet) {
+      return engine.state.facetSet[facetId].currentValues;
+    }
+    if (
+      engine.state.categoryFacetSet &&
+      facetId in engine.state.categoryFacetSet
+    ) {
+      return engine.state.categoryFacetSet[facetId]!.request.currentValues;
+    }
+    if (
+      engine.state.numericFacetSet &&
+      facetId in engine.state.numericFacetSet
+    ) {
+      return engine.state.numericFacetSet[facetId].currentValues;
+    }
+    if (engine.state.dateFacetSet && facetId in engine.state.dateFacetSet) {
+      return engine.state.dateFacetSet[facetId].currentValues;
+    }
+    return null;
+  };
+
+  const updateLastObservedValues = () => {
+    let anyParentWasUpdated = false;
+    Object.keys(lastObservedValuesPerParentFacet).forEach((parentFacetId) => {
+      const stringifiedValues = JSON.stringify(
+        getFacetValuesById(parentFacetId)
+      );
+      if (
+        stringifiedValues === lastObservedValuesPerParentFacet[parentFacetId]
+      ) {
+        return;
+      }
+      lastObservedValuesPerParentFacet[parentFacetId] = stringifiedValues;
+      anyParentWasUpdated = true;
+    });
+    return anyParentWasUpdated;
+  };
+
+  const areDependenciesMet = () => {
+    return props.dependencies.some((dependency) => {
+      if (!isFacetEnabled(dependency.parentFacetId)) {
+        return false;
+      }
+      const values = getFacetValuesById(dependency.parentFacetId);
+      if (values === null) {
+        return false;
+      }
+      return dependency.isDependencyMet(values as AnyFacetValue[]);
+    });
+  };
+
+  const thawFacets = () => {
+    if (engine.state.facetSet) {
+      Object.entries(engine.state.facetSet).forEach(
+        ([facetId, request]) =>
+          request.freezeCurrentValues &&
+          engine.dispatch(
+            updateFreezeCurrentValues({facetId, freezeCurrentValues: false})
+          )
+      );
+    }
+  };
```

Comment:
Without thawing facets, some facets won't fetch new values upon being enabled or upon another facet getting disabled/getting its values cleared.

Reply:
why is thawing only for specific facets?

Reply:
because they're the only ones that can (and need to) be thawed